test/apps/iodemo/io_demo.cc

```diff
@@ -584,68 +604,96 @@ class DemoClient : public P2pDemoCommon {
-            if (!conn[i]) {
-                LOG << "Connect to server ["
-                    << opts().servers[i]
-                    << "] Failed!";
-                for (size_t j = 0; j < i; j++) {
-                    delete conn[j];
+        if (_conn.empty()) {
+            _conn.resize(opts().servers.size());
+        }
+        for (size_t i = 0; i < _conn.size(); i++) {
+            if (_conn[i].conn == NULL) {
+                _conn[i].conn = connect(opts().servers[i], i + 1);
+                if (!_conn[i].conn) {
+                    LOG << "Connect to server ["
+                        << opts().servers[i]
+                        << "] Failed!";
+                } else {
+                    _num_connected++;
                 }
-                return false;
             }
         }
-
+        if (_num_connected == 0) {
+            return false;
+        }
         _status = OK;
 
         // TODO reset these values by canceling requests
         _num_sent      = 0;
         _num_completed = 0;
 
-        double prev_time     = get_time();
-        long total_iter      = 0;
-        long total_prev_iter = 0;
+        double prev_time           = get_time();
+        double prev_reconnect_time = get_time();
+        long total_iter            = 0;
+        long total_prev_iter       = 0;
         std::vector<op_info_t> info;
 
         for (int i = 0; i < IO_COMP; ++i) {
             op_info_t op_info = {static_cast<io_op_t>(i), 0, 0};
             info.push_back(op_info);
         }
 
-        while ((total_iter < opts().iter_count) && (_status == OK)) {
+        while ((total_iter < opts().iter_count)) {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
             if (!wait_for_responses(opts().window_size - 1)) {
                 break;
             }
 
-            size_t conn_num = IoDemoRandom::rand(0, conn.size() - 1);
-            io_op_t op      = get_op();
+            if (_num_connected < _conn.size()) {
+                double curr_time = get_time();
+                if (curr_time >= (prev_reconnect_time + 3.0)) {
+                    prev_reconnect_time = curr_time;
+                    for (size_t i = 0; i < _conn.size(); i++) {
+                        if (_conn[i].conn == NULL) {
+                            _conn[i].conn = connect(opts().servers[i], i + 1);
+                            if (_conn[i].conn) {
+                                _num_connected++;
+                            }
+                        }
+                    }
+                }
+            }
+            if (_num_connected == 0) {
+                LOG << "Are servers are disconnected";;
+                _status = ERROR;
+                break;
+            }
+            size_t conn_index;
+            while (1) {
+                conn_index = IoDemoRandom::rand(0, _conn.size() - 1);
+                if (_conn[conn_index].conn != NULL) {
+                    break;
+                }
+            }
```

Comment:
pls don't throw dice, keep alive server indexes somewhere

Reply:
@evgeny-leksikov NULL indicates not connected, other values indicate connected. This is according to the return value of connect() function. What's the benifits to  add another copy of variables to keep server connection status?

Reply:
to avoid infinite loop, is there any little chance what all servers are not connected? let's say local NIC is down, for example

Reply:
@evgeny-leksikov I don't think there is this case here. We have ``_num_connected`` checking before this code.
```            if (_num_connected == 0) {
                LOG << "Are servers are disconnected";;
                _status = ERROR;
                break;
            }```

Reply:
anyway better avoid operations with unpredictable complexity on data path